File: P2.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 19 10:26:26 2021

@author: jgdch
"""

#Space Complexity:O(H)
#Time Complexity:O(N)
#Passed all LC test cases
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: TreeNode):
        
        def preorder(r, curr_num):
            nonlocal sum_rl
            if r:
                curr_num = curr_num * 10 + r.val
                if not (r.left or r.right):
                    sum_rl += curr_num
                    
                preorder(r.left, curr_num)
                preorder(r.right, curr_num) 
        
        sum_rl= 0
        preorder(root, 0)
        return sum_rl        

File: p2.java
// Time Complexity : O(n)
// Space Complexity :O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //Global variable sum which is the result
    int sum = 0;
    public int sumNumbers(TreeNode root) {
        //Passing root node and local variable sum
        helper(root, 0);
        return sum;
    }

    public void helper(TreeNode root, int rs){

//Base condiiton
// When node is null, return
        if(root == null) return;

        //Getting the local variable value of sum
        rs = rs*10 + root.val;

        //If it is a leaf node, add it to the global variable sum
        if(root.left == null && root.right == null){
            sum = sum + rs;
            return;
        }

        
//Left node recursion
        helper(root.left, rs);
//Right node recurison
        helper(root.right, rs);
    }
}